File: QF_Cpp/test/UserEvent/autocode/UserEventTestImpl.cpp
//-*- Mode: C++; -*-
//===========================================================================
// This software contains Caltech/JPL confidential information.
//
// Copyright 2009-2011, by the California Institute of Technology.
// ALL RIGHTS RESERVED. United States Government Sponsorship Acknowledged.
// Any commercial use must be negotiated with the Office of Technology
// Transfer at the California Institute of Technology.
//
// This software may be subject to US export control laws and
// regulations. By accepting this document, the user agrees to comply
// with all applicable U.S. export laws and regulations, including the
// International Traffic and Arms Regulations, 22 C.F.R. 120-130 and the
// Export Administration Regulations, 15 C.F.R. 730-744. User has the
// responsibility to obtain export licenses, or other export authority as
// may be required before exporting such information to foreign countries
// or providing access to foreign persons.
//===========================================================================
//
//       File: UserEventTestImpl.cpp
// Created on: 30-Jul-2011 10:20:49
//     Author: scheng@jpl.nasa.gov
//
// This file was stubbed by the JPL StateChart Autocoders, which converts UML
// Statecharts, in XML, to a C++ variant of Miro Samek's Quantum Framework.
//===========================================================================
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <log_event.h>
#include <qf_port.h>
#include <qassert.h>
#include <UserEventImpl.h>
#include <UserEventTestImpl.h>
#include <StatechartSignals.h>

/**
 * Default constructor.
 */
UserEventTestImpl::UserEventTestImpl () {
    strcpy(this->machineName, "UserEventTest");
    strcpy(this->var, "None");

    AttributeMapper::init(this);
}

/**
 * Required virtual destructor.
 */
UserEventTestImpl::~UserEventTestImpl () {
    AttributeMapper::clean(this);
}

void UserEventTestImpl::setQActive (QActive* active) {
    m_active = active;
}


////////////////////////////////////////////
// Action and guard implementation methods
////////////////////////////////////////////

bool UserEventTestImpl::isValue (const char* arg1, double arg2) {
	// parse string val as double
	double varVal;
	sscanf(arg1, "%lf", &varVal);
    bool rv = (varVal == arg2);
    printf("%s.isValue() == %s\n", this->machineName, AttributeMapper::booltostr(rv));
    return rv;
}

void UserEventTestImpl::noUserEventOnInit (const char* arg1, const char* arg2) {
    printf("%s.noUserEventOnInit() 'var' originally '%s', setting to '%s'\n",
    		this->machineName, arg1, arg2);
	strcpy(this->var, arg2);
}

void UserEventTestImpl::noUserEvent (const char* arg1, QEvent const* e) {
    printf("%s.noUserEvent() received var value '%s' and event signal '%hd'\n",
    		this->machineName, arg1, e->sig);

    if (AttributeMapper::get(this, "FUNCTIONALITY")) {
    	// show that there is no data!
		DataEvent const* de = static_cast<DataEvent const*>(e);
    	printf("%s.noUserEvent() no access to user event on entry/exit: '%hX'\n",
    			this->machineName, ((uint8_t*) de)[3]);
    }
}

void UserEventTestImpl::userEventOnTrans (QEvent const* e, const char* arg2, double arg3) {
    printf("%s.userEventOnTrans() received event signal '%hd', var value '%s', and literal '%lf'\n",
    		this->machineName, e->sig, arg2, arg3);

    if (AttributeMapper::get(this, "FUNCTIONALITY")) {
		// transfer e data to var, and save literal as new e data
		DataEvent const* de = static_cast<DataEvent const*>(e);
		strncpy(this->var, de->data, MAX_DATA_SIZE);
		// publish new data in e as test Ack event
		DataEvent* newDE = Q_NEW(DataEvent, TEST_ACK_EV1);
		sprintf(newDE->data, "%lf", arg3);
		QF::publish(newDE);
	} else {
		// set var to literal
		sprintf(this->var, "%lf", arg3);
	}
}

void UserEventTestImpl::userEventAvailable (const char* arg1, QEvent const* e, const char* arg3) {
    printf("%s.userEventAvailable() received string '%s', event signal '%d', and var value '%s'\n",
    		this->machineName, arg1, e->sig, arg3);

    if (AttributeMapper::get(this, "FUNCTIONALITY")) {
		// transfer e data back to var
		DataEvent const* de = static_cast<DataEvent const*>(e);
		strncpy(this->var, de->data, MAX_DATA_SIZE);
    }
}

File: QF_Py/test/UserEvent/autocode/UserEventTestImpl.py
#
# Copyright 2009, 2010 California Institute of Technology.
# ALL RIGHTS RESERVED. U.S. Government Sponsorship acknowledged.
#
"""
File: UserEventTestImpl.py

Date Created:  02-Jun-2010 11:31:49
Created By:    scheng

Python custom-implementation class for functions referenced in
the UserEventTest Statechart model.
"""
# Python imports here
import sys
import logging

# QF imports here
from qf import framework
from qf import event

import UserEventImpl

# Module globals initialized here
LOGGER = logging.getLogger('UserEventTestLogger')


def printf(format, *args):
    # sys.stdout.write(format % args)
    LOGGER.info(format % args)


class UserEventTestImpl(object):
    """
    UserEventTest state machine implementation object.
    """
    def __init__(self, machine_name="UserEventTest"):
        """
        Constructor
        """
        self.__machine_name = machine_name
        self.fr = framework.QF.getInstance()
        self.activeObjects = self.fr.getActiveDict()
        #
        print "*Accessible: IMPL['UserEventTest']"
        self.set("FUNCTIONALITY", False)
        #
        self.var = None

    def set(self, attr, value=True):
        """
        Sets the boolean value of an attribute, default to True
        """
        if type(value) == bool:
            self.__dict__['_'+attr] = value
            print attr, "set to", value
        else:
            print 'Value must be a bool, either "True" or "False" (no quotes)!'

    def get(self, attrname):
        """
        Gets the boolean value of an attribute
        """
        return self.__dict__['_'+attrname]

    def clear(self, attrname):
        """
        Clears the value of an attribute to False
        """
        self.__dict__['_'+attrname] = False

    def isValue(self, arg1, arg2):
        """
        Implementation Guard method for isValue().
        """
        print "%s.isValue() == %s" % (self.__machine_name, arg1 == arg2)
        return arg1 == arg2

    def noUserEventOnInit(self, arg1, arg2):
        """
        Implementation Action method for noUserEventOnInit().
        """
        self.var = arg2
        print "%s.noUserEventOnInit() 'var' originally '%s', setting to '%s'"\
                % (self.__machine_name, arg1, arg2)

    def noUserEvent(self, arg1, e):
        """
        Implementation Action method for noUserEvent().
        The triggering event is supplied as a dictionary object 'e', so member
        data "data1" of the original Event object is accessible as e['data1'] .
        """
        print "%s.noUserEvent() received var value '%s' and event signal '%s'"\
                % (self.__machine_name, arg1, e['sType'])

        if self.get("FUNCTIONALITY"):
            print "%s.noUserEvent() no access to user event on entry/exit: '0'"\
                    % self.__machine_name

    def userEventOnTrans(self, e, arg2, arg3):
        """
        Implementation Action method for userEventOnTrans().
        The triggering event is supplied as a dictionary object 'e', so member
        data "data1" of the original Event object is accessible as e['data1'] .
        """
        print "%s.userEventOnTrans() received event signal '%s', var value '%s', and literal '%s'"\
                % (self.__machine_name, e['sType'], arg2, arg3)

        if self.get("FUNCTIONALITY"):
            self.var = e['data']
            self.fr.publish(UserEventImpl.DataEvent("TEST_ACK_EV1", arg3))
        else:
            self.var = arg3


    def userEventAvailable(self, arg1, e, arg3):
        """
        Implementation Action method for userEventAvailable().
        The triggering event is supplied as a dictionary object 'e', so member
        data "data1" of the original Event object is accessible as e['data1'] .
        """
        print "%s.userEventAvailable() received string '%s', event signal '%s', and var value '%s'"\
                % (self.__machine_name, arg1, e['sType'], arg3)

        if self.get("FUNCTIONALITY"):
            self.var = e['data']
